### PR TITLE
defend against TryPeek command returning unexpected value

### PR DESCRIPTION
**Pull request by adamralph**
alternative to https://github.com/Particular/NServiceBus.SqlServer/pull/846

### CODE REVIEW

**Review by andreasohlund — APPROVED** [review]

File: src/NServiceBus.Transport.SqlServer/Queuing/TableBasedQueue.cs [hunk 1]
```diff
@@ -37,7 +37,20 @@ public virtual async Task<int> TryPeek(SqlConnection connection, SqlTransaction
                 CommandTimeout = timeoutInSeconds ?? 30
             })
             {
-                var numberOfMessages = (int)await command.ExecuteScalarAsync(cancellationToken).ConfigureAwait(false);
+                var result = await command.ExecuteScalarAsync(cancellationToken).ConfigureAwait(false);
+
+                if (!(result is int numberOfMessages))
```
Comment (ramonsmits) [inline src/NServiceBus.Transport.SqlServer/Queuing/TableBasedQueue.cs:42]: How amount the other `ExecuteScalarAsync` calls?

File: src/NServiceBus.Transport.SqlServer/Queuing/TableBasedQueue.cs [hunk 2]
```diff
@@ -37,7 +37,20 @@ public virtual async Task<int> TryPeek(SqlConnection connection, SqlTransaction
...
+                {
+                    if (result is null)
+                    {
+                        throw new Exception("The peek command returned a null value.");
+                    }
+                    else
```
Comment (ramonsmits) [inline src/NServiceBus.Transport.SqlServer/Queuing/TableBasedQueue.cs:48]: I don't think the else could ever happen for this query but would make sense if this validation would be extracted to a method for every `ExecuteScalarAsync` call.

File: src/NServiceBus.Transport.SqlServer/Queuing/TableBasedQueue.cs (hunk 1 above)
Comment (adamralph) [reply src/NServiceBus.Transport.SqlServer/Queuing/TableBasedQueue.cs:42]: 🤷 I guess they should also be fixed.

File: src/NServiceBus.Transport.SqlServer/Queuing/TableBasedQueue.cs (hunk 2 above)
Comment (adamralph) [reply src/NServiceBus.Transport.SqlServer/Queuing/TableBasedQueue.cs:48]: That probably makes sense. These calls should not start throwing null ref's etc. when DB connectivity problems yield unexpected results or when mistakes have been made elsewhere (e.g. bad injected SQL).

**Comment by DavidBoike** [comment]
There's currently a SQL Transport enhancement release task force running. We decided to close this in favor of https://github.com/Particular/NServiceBus.SqlServer/pull/846.

File: src/NServiceBus.Transport.SqlServer/Queuing/TableBasedQueue.cs (hunk 1 above)
Comment (adamralph) [reply src/NServiceBus.Transport.SqlServer/Queuing/TableBasedQueue.cs:42]: see https://github.com/Particular/NServiceBus.SqlServer/pull/851

File: src/NServiceBus.Transport.SqlServer/Queuing/TableBasedQueue.cs (hunk 2 above)
Comment (adamralph) [reply src/NServiceBus.Transport.SqlServer/Queuing/TableBasedQueue.cs:48]: see https://github.com/Particular/NServiceBus.SqlServer/pull/851

**Pull request closed without merging**